libbpf-rs/src/object.rs

```diff
@@ -284,12 +344,87 @@ pub struct Object {
 }
 
 impl Object {
-    fn new(ptr: *mut libbpf_sys::bpf_object) -> Self {
-        Object {
+    fn new(ptr: *mut libbpf_sys::bpf_object) -> Result<Self> {
+        let mut obj = Object {
             ptr,
             maps: HashMap::new(),
             progs: HashMap::new(),
+        };
+
+        // Populate obj.maps
+        let mut map: *mut libbpf_sys::bpf_map = std::ptr::null_mut();
+        loop {
+            // Get the pointer to the next BPF map
+            let next_ptr = unsafe { libbpf_sys::bpf_map__next(map, obj.ptr) };
+            if next_ptr.is_null() {
+                break;
+            }
+
+            // Get the map name
+            let name = unsafe { libbpf_sys::bpf_map__name(next_ptr) };
+            let err = unsafe { libbpf_sys::libbpf_get_error(name as *const _) };
+            if err != 0 {
+                return Err(Error::System(err as i32));
+            }
+            let name = util::c_ptr_to_string(name)?;
+
+            // Get the map def
+            // bpf_map__def can return null but only if it's passed a null.
+            // We already know next_ptr is not null.
+            let def = unsafe { ptr::read(libbpf_sys::bpf_map__def(next_ptr)) };
+
+            // Get the map fd
+            let fd = unsafe { libbpf_sys::bpf_map__fd(next_ptr) };
+            if fd < 0 {
+                return Err(Error::System(errno::errno()));
```

Comment:
I think the existing code is wrong. libbpf returns the negative errno on errors: https://github.com/libbpf/libbpf/blob/c5389a965bc3f19e07b1ee161092fc227e364e94/src/libbpf.c#L9382-L9385

```suggestion
                return Err(Error::System(-fd));
```